cmd/global-query/pkg/distributed/query.go

```diff
@@ -37,6 +39,18 @@ func NewQueryRunner(resolver hosts.Resolver, querier Querier, opts ...QueryOptio
 	return
 }
 
+// SetResultReceivedFn registers a callback to be executed for every results that is
+// read off the results channel
+func (q *QueryRunner) SetResultReceivedFn(f func(*results.Result) error) *QueryRunner {
+	q.onResult = f
+	return q
+}
+
+// ResultReceived calls the result callback with res and
+func (q *QueryRunner) ResultReceived(res *results.Result) error {
+	return q.onResult(res)
```

Comment:
This isn't guarded against `q.onResult == nil` (I known you're checking below, but it would probably be safer to guard against it here to avoid some unexpected crash just because some code got reshuffled later)...